src/sctools/fastq.py

```diff
@@ -160,41 +160,88 @@ def __iter__(self):
 EmbeddedBarcode = namedtuple('Tag', ['start', 'end', 'sequence_tag', 'quality_tag'])
 
 
+def extract_barcode(record, embedded_barcode):
+    """extracts barcodes from a fastq record according to an EmbeddedBarcode object.
+
+    :param FastqRecord record: record to extract from
+    :param EmbeddedBarcode embedded_barcode: defines the barcode start and end and the names of the quality
+      and sequence tags
+    :return tuple: (sequence tag name, tag content (nucleotides), tag type (Z))
+    :return tuple: (quality tag name, tag content (qualities), tag type (Z))
+    """
+    seq = record.sequence[embedded_barcode.start:embedded_barcode.end]
+    qual = record.quality[embedded_barcode.start:embedded_barcode.end]
+    return (embedded_barcode.sequence_tag, seq, 'Z'), (embedded_barcode.quality_tag, qual, 'Z')
+
+
 class EmbeddedBarcodeGenerator(Reader):
 
-    def __init__(self, embedded_barcodes, *args, **kwargs):
+    def __init__(self, fastq_files, embedded_barcodes, *args, **kwargs):
         """
         parse fastq files for barcodes(s) defined by EmbeddedBarcode objects, producing a generator
         that yields extracted barcode objects in a form that is consumable by pysam's bam and sam
         set_tag methods.
 
         :param [EmbeddedBarcode] embedded_barcodes: list of tag objects defining start and end of
           the sequence containing the tag, plus the string quality and sequence tags
-        :param list|str files: (default sys.stdin) fastq file or list of files to be read.
+        :param list|str fastq_files: (default sys.stdin) fastq file or list of files to be read.
         :param mode: (Default 'r') will return string objects. Change to 'rb' to
           return bytes objects.
         """
-        super().__init__(*args, **kwargs)
+        super().__init__(files=fastq_files, *args, **kwargs)
         self.embedded_barcodes = embedded_barcodes
 
     def __iter__(self):
         """iterates over barcodes extracted from fastq"""
         for record in super().__iter__():  # iterates records; we extract barcodes.
             barcodes = []
             for barcode in self.embedded_barcodes:
-                barcodes.extend(self.extract_barcode(record, barcode))
+                barcodes.extend(extract_barcode(record, barcode))
             yield barcodes
 
-    @staticmethod
-    def extract_barcode(record, eb):
-        """extracts barcodes from a fastq record according to an EmbeddedBarcode object.
-
-        :param FastqRecord record: record to extract from
-        :param EmbeddedBarcode eb: defines the barcode start and end and the names of the quality
-          and sequence tags
-        :return tuple: (sequence tag name, tag content (nucleotides), tag type (Z))
-        :return tuple: (quality tag name, tag content (qualities), tag type (Z))
+
+class BarcodeGeneratorWithCorrectedCellBarcodes(Reader):
```

Comment:
Why have a separate class for handing corrected cell barcodes instead of modifying `EmbeddedBarcodeGenerator`?

Reply:
So, I initially subclassed `EmbeddedBarcodeGenerator` instead of `Reader`, but that did weird stuff to `__init__`, which I would also have to overload. Since I was basically overloading all the methods, it made more sense to do a parallel subclassing of reader. 

I created a separate method because it didn't seem worth it to generalize to `CorrectedEmbeddedBarcodeGenerator` at the time; I would need to pair whitelists with barcodes in a separate class, then find a way to bring in barcodes with and without construction. This should be possible, but I feel I need to focus on creating a draft pipeline for the quarterly demo. I'm happy to make a tech debt ticket to combine these classes If I get all the pre-demo work done with time to spare. (?)

What do you think?

Reply:
Ok, thanks for the explanation! I think adding a follow-up ticket is fine

Reply:
Ok. In that case, I will leave the follow-up branch open -- there is potential for some debugging to be needed when I apply to it to wdl, and because of how Pypi works, I need to  increment the version each time I push it up. 

We can revisit merging the PR when all of the WDL coding is completed.

Reply:
Ticket made.